src/random.rs

```diff
@@ -0,0 +1,722 @@
+/// random number generation.
+pub fn key(seed: u64) -> Array {
+    unsafe { Array::from_ptr(mlx_sys::mlx_random_key(seed)) }
+}
+
+/// Split a PRNG key into two keys and return a tuple.
+#[default_device]
+pub fn split_device(key: &Array, stream: StreamOrDevice) -> (Array, Array) {
+    let keys = unsafe {
+        Array::from_ptr(mlx_sys::mlx_random_split_equal_parts(
+            key.as_ptr(),
+            2,
+            stream.as_ptr(),
+        ))
+    };
+
+    (keys.index(0), keys.index(1))
+}
+
+/// Generate uniformly distributed random numbers with a given interval (`lower_bound:` and `upper_bound:`).
+///
+/// The values are sampled uniformly in the range.  An optional shape can be used to broadcast into
+/// a larger array.  An optional `key` can be specified to control the PRNG.
+///
+/// ```rust
+/// let key = mlx_rs::random::key(0);
+///
+/// // create an array of shape `[50]` type Float values in the range `0 ..< 10`
+/// let array = mlx_rs::random::uniform::<_, f32>(0, 10, &[50], &key);
+///
+/// // same, but in range `0.5 ..< 1`
+/// let array = mlx_rs::random::uniform::<_, f32>(0.5f32, 1f32, &[50], &key);
+/// ```
+#[default_device]
+pub fn uniform_device<'a, E: Into<Array>, T: ArrayElement>(
+    lower_bound: E,
+    upper_bound: E,
+    shape: &[i32],
+    key: impl Into<Option<&'a Array>>,
+    stream: StreamOrDevice,
+) -> Result<Array, Exception> {
+    let lb: Array = lower_bound.into();
+    let ub: Array = upper_bound.into();
+
+    let key = key.into().map_or_else(
+        || {
+            let mut state = state().lock().unwrap();
+            state.next()
+        },
+        |key| key.clone(),
+    );
+
+    unsafe {
+        let c_array = try_catch_c_ptr_expr! {
+            mlx_sys::mlx_random_uniform(
+                lb.as_ptr(),
+                ub.as_ptr(),
+                shape.as_ptr(),
+                shape.len(),
+                T::DTYPE.into(),
+                key.as_ptr(),
+                stream.as_ptr(),
+            )
+        };
+        Ok(Array::from_ptr(c_array))
+    }
+}
+
+/// Generate normally distributed random numbers.
+///
+/// Generate an array of random numbers using the optional shape. The result
+/// will be of the given `T`. `T` must be a floating point type.
+///
+/// ```rust
+/// let key = mlx_rs::random::key(0);
+///
+/// // generate a single Float with normal distribution
+/// let value = mlx_rs::random::normal::<f32>(None, None, None, &key).unwrap().item::<f32>();
+///
+/// // generate an array of Float with normal distribution in shape [10, 5]
+/// let array = mlx_rs::random::normal::<f32>(&[10, 5][..], None, None, &key);
+/// ```
+///
+/// # Params
+///  - shape: shape of the output, if `None` a single value is returned
+///  - loc: mean of the distribution, default is `0.0`
+///  - scale: standard deviation of the distribution, default is `1.0`
+///  - key: PRNG key
+#[default_device]
+pub fn normal_device<'a, T: ArrayElement>(
+    shape: impl Into<Option<&'a [i32]>>,
```

Comment:
This would accept `&[i32; N]` as args
```suggestion
    shape: impl IntoOption<&'a [i32]>,
```